src/lib/utils/__tests__/mergeSearchParameters-test.ts

```diff
@@ -0,0 +1,370 @@
+    const actual = merge(
+      algoliasearchHelper.SearchParameters.make({
+        // Inherit
+        hitsPerPage: 2,
+        attributesToSnippet: ['description'],
+        // Overridden
+        query: 'Samsung',
+        attributesToHighlight: ['name'],
+      }),
+      algoliasearchHelper.SearchParameters.make({
+        // hitsPerPage: 2,
+        // attributesToSnippet: ['description'],
+        query: 'Apple',
+        attributesToHighlight: ['name', 'author'],
+      }),
+      algoliasearchHelper.SearchParameters.make({
+        // hitsPerPage: 2,
+        // attributesToSnippet: ['description'],
+        // attributesToHighlight: ['name', 'author'],
+        query: 'Apple iPhone',
+        distinct: true,
+      })
+    );
+
+    expect(actual).toEqual(
+      algoliasearchHelper.SearchParameters.make({
+        hitsPerPage: 2,
+        attributesToSnippet: ['description'],
+        attributesToHighlight: ['name', 'author'],
+        query: 'Apple iPhone',
+        distinct: true,
+      })
+    );
+  });
+
+  it('merges `facets` parameters', () => {
+    const actual = merge(
+      algoliasearchHelper.SearchParameters.make({
+        facets: ['brand'],
+      }),
+      algoliasearchHelper.SearchParameters.make({
+        facets: ['categories'],
+      }),
+      algoliasearchHelper.SearchParameters.make({
+        facets: ['brand', 'colors'],
+      })
+    );
+
+    expect(actual).toEqual(
+      algoliasearchHelper.SearchParameters.make({
+        facets: ['brand', 'categories', 'colors'],
+      })
+    );
+  });
+
+  it('merges `disjunctiveFacets` parameters', () => {
+    const actual = merge(
+      algoliasearchHelper.SearchParameters.make({
+        disjunctiveFacets: ['brand'],
+      }),
+      algoliasearchHelper.SearchParameters.make({
+        disjunctiveFacets: ['categories'],
+      }),
+      algoliasearchHelper.SearchParameters.make({
+        disjunctiveFacets: ['brand', 'colors'],
+      })
+    );
+
+    expect(actual).toEqual(
+      algoliasearchHelper.SearchParameters.make({
+        disjunctiveFacets: ['brand', 'categories', 'colors'],
+      })
+    );
+  });
+
+  it('merges `facetsRefinements` parameters', () => {
+    const actual = merge(
+      algoliasearchHelper.SearchParameters.make({
+        facets: ['brand'],
+        facetsRefinements: {
+          brand: ['Samsung'],
+        },
+      }),
+      algoliasearchHelper.SearchParameters.make({
+        facets: ['categories'],
+        facetsRefinements: {
+          categories: ['TVs'],
+        },
+      }),
+      algoliasearchHelper.SearchParameters.make({
+        facets: ['brand', 'colors'],
+        facetsRefinements: {
+          brand: ['Apple'],
+          colors: ['Red'],
+        },
+      })
+    );
+
+    expect(actual).toEqual(
+      algoliasearchHelper.SearchParameters.make({
+        facets: ['brand', 'categories', 'colors'],
+        facetsRefinements: {
+          brand: ['Apple'],
```

Comment:
Are there any cases where we wouldn't want `Samsung` to be overridden by `Apple` but rather appended? (same question for other objects containing arrays)

Reply:
@Haroenv had the same question. The reasoning behind this implementation is to keep the behavior consistent between primitives and managed parameters. Applied to widgets it helps to understand:

```js
search.addWidgets([
  index({ indexName: "topLevel" }).addWidgets([
    // Top level searchBox
    searchBox(),
    index({ indexName: "subLevel" }).addWidgets([
      // Sub level searchBox
      searchBox()
    ])
  ])
]);
```

The two widgets `search box` are independent from each other. Each widget **controls** its part of the state where it is mounted. The sub-level `searchBox` does not inherit from the top-level one, it overrides it. We apply the same rule to the managed parameters that are on the **same attribute**. Note that this rule only applies to managed parameters shared across multiple indices. Within the same index both widgets can read/write on the internal state.

With React InstantSearch I've never seen a use case where multiple widgets inherit from each other within a multi-index context. The behavior of React InstantSeach is the same as the one described above. I agree that it's a tradeoff but to it's the behavior that makes sense (IMO) with the idea of controlled/uncontrolled widgets.